Support live disk resizing for RAW images

tpressure opened the pull request:
## Summary

This pull request introduces live disk resizing support for RAW images. This enhancement enables dynamic adjustment of disks while the guest is running. This is part of our [cloud-hypervisor openstack enablement](https://github.com/cyberus-technology/openstack-nova/tree/gardenlinux).

## Testing
- [x] Ran our [libvirt test suite](https://github.com/cyberus-technology/libvirt-tests).
- [x] Add integration test

tpressure committed:
ch-remote: support live disk resizing

Support disk resizing via ch-remote and REST api.

On-behalf-of: SAP thomas.prescher@sap.com
Signed-off-by: Thomas Prescher <thomas.prescher@cyberus-technology.de>

tpressure committed:
block: virtio-devices: make disk_nsectors a shared atomic

This change is a prerequisite for live disk resizing. Before this
commit, the epoll-handler threads just got a copy of the sector
size which we cannot update during runtime.

On-behalf-of: SAP thomas.prescher@sap.com
Signed-off-by: Thomas Prescher <thomas.prescher@cyberus-technology.de>

tpressure committed:
vmm: disk resize infrastructure

Add basic infrastructure so resize events are
propagated to the underlying disk implementation.

On-behalf-of: SAP thomas.prescher@sap.com
Signed-off-by: Thomas Prescher <thomas.prescher@cyberus-technology.de>

tpressure committed:
block: raw_async: implement disk resizing

Support for resize events for raw_async disks.

On-behalf-of: SAP thomas.prescher@sap.com
Signed-off-by: Thomas Prescher <thomas.prescher@cyberus-technology.de>

rbradford commented on vmm/src/api/mod.rs:
```diff
@@ -1135,6 +1147,44 @@ impl ApiAction for VmResize {
+    ) -> ApiRequest {
+        Box::new(move |vmm| {
+            info!("API request event: VmResizeDisk {:?}", resize_disk_data);
+            println!("xxxxxx");
```
Testing code? :-)

tpressure replied:
ouch. I somehow missed to remove this ;)

rbradford commented on src/bin/ch-remote.rs:
```diff
@@ -1022,6 +1050,20 @@ fn get_cli_commands_sorted() -> Box<[Command]> {
                     .num_args(1),
             ),
+        Command::new("resize-disk")
+            .about("grows/shrinks an attached disk")
```
Maybe this should be `.help()`

rbradford replied:
My mistake - I was reading the wrong context. I think maybe it should be "Resize attached disk"

tpressure committed:
docs: update vm.resize-disk

On-behalf-of: SAP thomas.prescher@sap.com
Signed-off-by: Thomas Prescher <thomas.prescher@cyberus-technology.de>

rbradford reviewed (commented):
I'm kinda impressed that Linux works with this! Will need an integration test.

rbradford commented on src/bin/ch-remote.rs:
```diff
@@ -1022,6 +1050,20 @@ fn get_cli_commands_sorted() -> Box<[Command]> {
                     .num_args(1),
             ),
+        Command::new("resize-disk")
+            .about("resize attached disk")
```
Is this capitalisation consistent with the other commands in this tool?

rbradford commented on virtio-devices/src/block.rs:
```diff
@@ -93,6 +93,14 @@ pub enum Error {
         /// The path of the disk image.
         path: PathBuf,
     },
+    #[error("disk image size is not a multiple of {}", SECTOR_SIZE)]
```
Capitalisation (I know Rust style has these as lower cases - but a lot of our code predates that establishment.)

rbradford reviewed (commented):
Thanks! I have no problems in principal with the feature/PR. Please try and add an integration test.

tpressure committed:
tests: add live disk resize test

This test verifies that we can grow and shrink
disk images during runtime.

On-behalf-of: SAP thomas.prescher@sap.com
Signed-off-by: Thomas Prescher <thomas.prescher@cyberus-technology.de>

tpressure commented on block/src/async_io.rs:
```diff
@@ -61,6 +69,8 @@ pub trait DiskFile: Send {
     fn topology(&mut self) -> DiskTopology {
         DiskTopology::default()
     }
+    fn resize(&mut self, size: u64) -> DiskFileResult<()>;
```
done

tpressure commented:
> I'm kinda impressed that Linux works with this! Will need an integration test.

@rbradford  test has been added.

tpressure commented:
> > @phip1611 you are absolutely right. Would it be ok to do this in a follow-up PR?
> 
> I think it would be sufficient to create a follow-up ticket. In fact, this also has some caveats. For example, next to this live resize functionality, in the static qcow2 case, the physical image can also grow transparently to the guest, if I'm not mistaken. No need to fix all of this here in this PR.

I've created https://github.com/cloud-hypervisor/cloud-hypervisor/issues/7569

rbradford reviewed (approved):
Just some nits.

tpressure commented on virtio-devices/src/block.rs:
```diff
@@ -95,6 +95,16 @@ pub enum Error {
+    #[error("Failed to resume vcpus")]
+    ResumeVcpus(#[source] MigratableError),
+    #[error("Failed signal config interrupt")]
+    SignalingConfigChange(#[source] io::Error),
```
There's also `FailedSignalingUsedQueue` but who's counting ;)

tpressure commented on vmm/src/device_manager.rs:
```diff
@@ -670,6 +670,10 @@ pub enum DeviceManagerError {
+
+    /// Disk resizing failed.
+    #[error("Disk resize error")]
+    DiskResizeError(#[source] virtio_devices::block::Error),
```
done

rbradford commented:
@tpressure Something came up when I was tasking an LLM with finding CH bugs - we accept (and return 204) when resizing down a raw file. Even though this can cause data loss. I think this is by design but I just want to check that was intended

tpressure commented:
> @tpressure Something came up when I was tasking an LLM with finding CH bugs - we accept (and return 204) when resizing down a raw file. Even though this can cause data loss. I think this is by design but I just want to check that was intended

Indeed this was intended. I tried to mimic the behavior of qemu here. Shrinking a disk of course requires guest support, as the guest must ensure that the data is no longer accessed. I can talk to our customers if this is really needed if we want to remove disk shirnking support.

rbradford commented:
> > @tpressure Something came up when I was tasking an LLM with finding CH bugs - we accept (and return 204) when resizing down a raw file. Even though this can cause data loss. I think this is by design but I just want to check that was intended
> 
> Indeed this was intended. I tried to mimic the behavior of qemu here. Shrinking a disk of course requires guest support, as the guest must ensure that the data is no longer accessed. I can talk to our customers if this is really needed if we want to remove disk shirnking support.

Thank you for your prompt response. I think this is fine.
